test/pyro/test_hmm.py

```diff
@@ -144,3 +146,58 @@ def test_gaussian_mrf_log_prob(init_shape, trans_shape, obs_shape, hidden_dim, o
     expected_log_prob = expected_dist.log_prob(data)
     assert_close(actual_log_prob, expected_log_prob, atol=1e-4, rtol=1e-4)
+
+
+@pytest.mark.parametrize("sample_shape", [(), (8,), (7, 6)], ids=str)
+@pytest.mark.parametrize("obs_dim", [3, 4])
+@pytest.mark.parametrize("hidden_dim", [1, 2])
+@pytest.mark.parametrize("init_shape,trans_matrix_shape,trans_dist_shape,obs_logits_shape,obs_dist_shape", [
+    ((), (6,), (), (), ()),
+    ((), (), (6,), (), ()),
+    ((), (), (), (6,), ()),
+    ((), (), (), (), (6,)),
+    ((5,), (5, 6), (5, 6), (5, 6), (5, 6)),
+], ids=str)
+def test_gaussian_discrete_mrf_shape(sample_shape, init_shape, trans_matrix_shape, trans_dist_shape,
+                                     obs_logits_shape, obs_dist_shape, obs_dim, hidden_dim):
+    init_dist = random_mvn(init_shape, hidden_dim)
+    trans_matrix = torch.randn(trans_matrix_shape + (hidden_dim, hidden_dim))
+    trans_dist = random_mvn(trans_dist_shape, hidden_dim)
+    obs_logits = torch.randn(obs_logits_shape + (obs_dim,))
+    obs_dist = random_mvn(obs_dist_shape + (obs_dim,), hidden_dim)
+
+    actual_dist = GaussianDiscreteMRF(init_dist, trans_matrix, trans_dist, obs_logits, obs_dist)
+
+    shape = broadcast_shape(init_shape + (1,),
+                            trans_matrix_shape,
+                            trans_dist_shape,
+                            obs_logits_shape,
+                            obs_dist_shape)
+    batch_shape, event_shape = shape[:-1], shape[-1:]
+    assert actual_dist.event_shape == event_shape
+    assert actual_dist.batch_shape == batch_shape
+
+    data = dist.Categorical(logits=obs_logits).expand(shape).sample(sample_shape)
+    actual_log_prob = actual_dist.log_prob(data)
+    assert actual_log_prob.shape == sample_shape + batch_shape
+
+
+@pytest.mark.parametrize("obs_dim", [2, 3])
+@pytest.mark.parametrize("hidden_dim", [1, 2])
+@pytest.mark.parametrize("num_steps", [2, 3, 4, 5])
+@pytest.mark.parametrize("batch_shape", [(), (5,), (3, 2)], ids=str)
+def test_gaussian_discrete_mrf_normalized(batch_shape, num_steps, hidden_dim, obs_dim):
+    init_dist = random_mvn(batch_shape, hidden_dim)
+    trans_matrix = torch.randn(batch_shape + (num_steps, hidden_dim, hidden_dim))
+    trans_dist = random_mvn(batch_shape + (num_steps,), hidden_dim)
+    obs_logits = torch.randn(batch_shape + (num_steps, obs_dim))
+    obs_dist = random_mvn(batch_shape + (num_steps, obs_dim), hidden_dim)
+    actual_dist = GaussianDiscreteMRF(init_dist, trans_matrix, trans_dist, obs_logits, obs_dist)
+
+    data = torch.tensor(list(itertools.product(*[list(range(obs_dim))] * num_steps)))
+    assert data.shape == (obs_dim ** num_steps, num_steps)
+    data = data.reshape(data.shape[:1] + (1,) * len(batch_shape) + data.shape[1:])
+    log_prob = actual_dist.log_prob(data)
+    assert log_prob.shape[1:] == batch_shape
+    log_total = log_prob.logsumexp(0)
+    assert log_total.abs().max() < 4, log_total
```

Comment:
`exp(4.0)` seems like a big error factor 🤔

Reply:
what happens if you make the moment matching more expensive, i.e. keep more components?

Reply:
Hmm, I'm not sure how to code that up. Another possible explanation is that the moment-matching approximation might be bad for random models (as in the test), but might be good for more plausible identifiable models. To test this I guess I could use plausible parameters above rather than random.